### DISCUSSION TITLE
what is difference between bpftrace and BCC?

### DISCUSSION THREAD

**Discussion by caneatcansleep** [bpftrace/bpftrace#2450] [Q&A]
Currently I know bpftrace is based on BCC, so does it mean bpftrace can do whatever BCC can do? Specifically, I would like to ask the following questions.
1. Does bpftrace support all bpf-helpers? Since BCC can use bpf-helpers such as bpf_l3_csum_replace(), bpf_tail_call() and bpf_redirect().
2. Does bpftrace support all ebpf-program types? Such as BPF_PROG_TYPE_SK_SKB, BPF_PROG_TYPE_SOCKET_FILTER and BPF_PROG_TYPE_XDP?
3. Does bpftrace support all ebpf-map types? Such as BPF_MAP_TYPE_SOCKMAP, BPF_MAP_TYPE_LRU_HASH?

[bpffeature.h](https://github.com/iovisor/bpftrace/blob/3cd853f81791a9305542ed6ddfeee10c8d088022/src/bpffeature.h) shows that bpftrace only supports several prog types and map types, but I am not sure.

**Comment by fbs** [comment]
bpftrace only supports whats builtin into the language. No arbitrary helper calls or map types
